### ISSUE TITLE
New Lint - Pointless mutation

### ISSUE THREAD

**Issue by kinnison** [rust-lang/rust-clippy#13977]
### What it does

The lint should detect situations in which newly created values are mutated and then immediately dropped during assignments.

### Advantage

Would spot situations such as the example below and warn the user that it is not acting as the author intended.

### Drawbacks

Deliberate discarding of values may end up being slightly less ergonomic in some corner cases.

### Example

This example is possibly a bit wordy, but is a simplified example of something encountered during a code review at work:

```rust
    struct Thingy {
        foo: Option<(u32, u32)>,
    }

    let mut thingy = Thingy { foo: None };
    thingy.foo = Some((0, 1));

    // This operation is effectively useless (no side-effects etc.)
    thingy.foo.unwrap().1 = 2;

    // What intuitively we'd expected to pass
    assert_eq!(thingy.foo.unwrap().1, 2);
```

The lint is clearly meant to trigger on the `thingy.foo.unwrap().1 = 2;` line and probably ought to suggest:

Could be written as:

```rust
    thingy.foo.as_mut().unwrap().1 = 2;
```

Though the exact details of the suggestion would likely tie the lint to only working on replacing the innards of `Option`s or similar.

**Comment by kinnison** [comment]
Some extra thoughts:

If the thing being unwrapped wasn't `Copy` then we'd get a use-after-move error anyway, so this only seems to apply in cases where the assignment is into an lvalue which was copied.

**Comment by GuillaumeGomez** [comment]
I think it's a valid case.

**Comment by kinnison** [comment]
If someone could guide me into how I might go about building such a lint (eg. pointing me at anything which might be similar) I'm prepared to have a go at a solution myself; though it'd be my first time in the clippy codebase.

**Comment by leonardo-m** [comment]
`thingy.foo.unwrap().1 = 2;`

Perhaps it's rustc that should warn for code like this, instead of Clippy?

**Comment by kinnison** [comment]
> `thingy.foo.unwrap().1 = 2;`
> 
> Perhaps it's rustc that should warn for code like this, instead of Clippy?

I'd be quite content if it were `rustc` itself which warned, rather than it being a Clippy extra, but I'm not really in a position to say which would be better.

**Comment by djc** [comment]
I just ran into something similar:

```rust
fn foo(mut foo: Foo) {
      foo.values.retain(|bar| bar.is_baz());
      .. // no more references to `foo.values` and `foo` is dropped here
}

struct Foo {
      values: Vec<Bar>,
}
```

This caused an issue that was a little tricky to find.
